### ISSUE TITLE
C# Grpc.Tools generates warning "Some expected protoc outputs were not generated"

### ISSUE THREAD

**Issue by libichong** [grpc/grpc#19816]
### What version of gRPC and what language are you using?
 I tried 1.21 and 1.22 version within C# project
 
### What operating system (Linux, Windows,...) and version?
 Windows 10
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 Microsoft (R) Build Engine version 15.7.177.53362 for .NET Framework
Visual Studio 2017 
 
### What did you expect to see?
 
 Surpress the warning or what condition can trigger the warning. 
 
### Anything else we should know about your project / environment?

```
d:\dbs\sh\mvdg\0801_025010\cmd\0\packages\Grpc.Tools.1.22.0\build\_protobuf\Google.Protobuf.Tools.targets(313,5): warning : Some expected protoc outputs were not generated. [d:\dbs\sh\mvdg\0801_025010\cmd\0\private\IndexGen\multimedia\VideoDownloader\VideoDownloaderServer\VideoDownloaderServer.csproj]
d:\dbs\sh\mvdg\0801_025010\cmd\0\packages\Grpc.Tools.1.22.0\build\_protobuf\Google.Protobuf.Tools.targets(313,5): warning :  [d:\dbs\sh\mvdg\0801_025010\cmd\0\private\IndexGen\multimedia\VideoDownloader\VideoDownloaderServer\VideoDownloaderServer.csproj]
d:\dbs\sh\mvdg\0801_025010\cmd\0\packages\Grpc.Tools.1.22.0\build\_protobuf\Google.Protobuf.Tools.targets(313,5): warning :     d:\dbs\sh\mvdg\0801_025010\\private\IndexGen\multimedia\VideoDownloader\VideoDownloaderServer\obj\amd64\VideoDownloaderProtocol.cs [d:\dbs\sh\mvdg\0801_025010\cmd\0\private\IndexGen\multimedia\VideoDownloader\VideoDownloaderServer\VideoDownloaderServer.csproj]
d:\dbs\sh\mvdg\0801_025010\cmd\0\packages\Grpc.Tools.1.22.0\build\_protobu
```

**Comment by jtattermusch** [comment]
I understand that the warning `Google.Protobuf.Tools.targets(313,5): warning : Some expected protoc outputs were not generated.` is the problem that you're seeing.
Could you provide a reproduction (a simple csproj project) that will trigger this warning? 
Without knowing the structure of your .csproj, it is unlikely that we will be able to reproduce this issue and help.

**Comment by amitwaisel** [comment]
@jtattermusch It happened to me as well when I include `proto` files to a dotnet-core `csproj` file, by placing:
```csproj
  <ItemGroup>
    <PackageReference Include="Google.Protobuf" Version="3.9.1" />
    <PackageReference Include="Grpc" Version="2.23.0" />
    <PackageReference Include="Grpc.Tools" Version="2.23.0">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
    </PackageReference>
    ...
  </ItemGroup>
  ...
  <ItemGroup>
    <ProtoBuf Include="..\rpc\proto\*.proto" />
  </ItemGroup>
```

In addition, I changed the `obj` directory location by placing
```csproj
  <PropertyGroup>
    <TargetFramework>netcoreapp2.2</TargetFramework>
    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
    <Platforms>AnyCPU;x64</Platforms>
    <AppendTargetFrameworkToOutputPath>true</AppendTargetFrameworkToOutputPath>
    <AppendTargetFrameworkToIntermediatePath>true</AppendTargetFrameworkToIntermediatePath>
  </PropertyGroup>

  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
    <OutputPath>$(SolutionDir)\..\bin\$(Platform)_$(Configuration)\</OutputPath>
    <IntermediateOutputPath>$(SolutionDir)\..\obj\$(Platform)_$(Configuration)\$(ProjectName)\</IntermediateOutputPath>
  </PropertyGroup>
  ...
```

This is the error message from the build output window:
```log
NuGet package restore failed. Please see Error List window for detailed warnings and errors.
1>------ Build started: Project: BlahCore, Configuration: Debug x64 ------
1>C:\Users\Blah\.nuget\packages\grpc.tools\2.23.0\build\_protobuf\Google.Protobuf.Tools.targets(313,5): warning : Some expected protoc outputs were not generated.
1>C:\Users\Blah\.nuget\packages\grpc.tools\2.23.0\build\_protobuf\Google.Protobuf.Tools.targets(313,5): warning : 
1>C:\Users\Blah\.nuget\packages\grpc.tools\2.23.0\build\_protobuf\Google.Protobuf.Tools.targets(313,5): warning :     C:\blah\src\\..\obj\x64_Debug\\netcoreapp2.2\Blah.cs
1>C:\Users\Blah\.nuget\packages\grpc.tools\2.23.0\build\_protobuf\Google.Protobuf.Tools.targets(313,5): warning :     C:\blah\src\\..\obj\x64_Debug\\netcoreapp2.2\BlahGrpc.cs
...
```

I hope it helps you to pinpoint the issue.

**Comment by joshlang** [comment]
We get the same warnings with this:

```
  <ItemGroup>
    <Protobuf Include="../AccountInfoService.Proto/**/*.proto" ProtoRoot="../AccountInfoService.Proto" GrpcServices="Server" OutputDir="Temp" CompileOutputs="false" />
  </ItemGroup>
```

However, everything works fine anyway.

**Comment by bddckr** [comment]
We're getting the warning, too. It does not give us any issues (the C# file is generated and put into our custom obj folder).

We're using
```xml
  <ItemGroup>
    <Protobuf Include="**/*.proto" GrpcServices="None" Access="Internal" />
    <Protobuf Update="**/*_service.proto" GrpcServices="Both" />
  </ItemGroup>
```
with a customized `IntermediateOutputPath`.

**Comment by PetruRitivoiu** [comment]
I get the same warnings. Here's my configuration

`<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Grpc" Version="2.27.0" />
    <PackageReference Include="Grpc.Core" Version="2.27.0" />
    <PackageReference Include="Grpc.Reflection" Version="2.27.0" />
    <PackageReference Include="Grpc.Tools" Version="2.27.0">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
  </ItemGroup>

  <ItemGroup>
    <Folder Include="models\" />
  </ItemGroup>


  <ItemGroup>
    <Protobuf Include="proto\*.proto" OutputDir="%(RelativePath)models/" />
  </ItemGroup>

</Project>`

`syntax = "proto3";

package dummy;

message DummyMessage {}

service DummyService {}`

`1>C:\Users\Petru_Ritivoiu\.nuget\packages\grpc.tools\2.27.0\build\_protobuf\Google.Protobuf.Tools.targets(313,5): warning : Some expected protoc outputs were not generated.
1>C:\Users\Petru_Ritivoiu\.nuget\packages\grpc.tools\2.27.0\build\_protobuf\Google.Protobuf.Tools.targets(313,5): warning : 
1>C:\Users\Petru_Ritivoiu\.nuget\packages\grpc.tools\2.27.0\build\_protobuf\Google.Protobuf.Tools.targets(313,5): warning :     models/Dummy.cs
1>C:\Users\Petru_Ritivoiu\.nuget\packages\grpc.tools\2.27.0\build\_protobuf\Google.Protobuf.Tools.targets(313,5): warning :     models/DummyGrpc.cs
1>C:\Users\Petru_Ritivoiu\.nuget\packages\grpc.tools\2.27.0\build\_protobuf\Google.Protobuf.Tools.targets(313,5): warning :     models/Dummy.cs
1>C:\Users\Petru_Ritivoiu\.nuget\packages\grpc.tools\2.27.0\build\_protobuf\Google.Protobuf.Tools.targets(313,5): warning :     models/DummyGrpc.cs
1>CSC : warning CS2002: Source file 'models\Dummy.cs' specified multiple times
1>CSC : warning CS2002: Source file 'models\DummyGrpc.cs' specified multiple times`

**Comment by 9ee1** [comment]
I can confirm this problem is still happening with `Grpc.Tools.2.27.0`. I am seeing similar behavior as everyone else. If I specify a custom intermediary output directory for my project in the project file, it errors out.

```
<Project Sdk="Microsoft.NET.Sdk.Web">
  <PropertyGroup>
    <AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
    <Nullable>enable</Nullable>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <Platforms>AnyCPU;x64;x86</Platforms>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
    <IntermediateOutputPath>../../../build/backend/debug-any-cpu/obj/</IntermediateOutputPath>
    <OutputPath>../../../build/backend/debug-any-cpu/bin/</OutputPath>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Grpc.AspNetCore" Version="2.28.0" />
    <PackageReference Include="Grpc.Reflection" Version="2.28.1" />
    <PackageReference Include="Microsoft.AspNetCore.Authentication.JwtBearer" Version="3.1.3" />
    <PackageReference Include="Microsoft.AspNetCore.SpaServices.Extensions" Version="3.1.3" />
    <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="3.1.3" />
    <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="3.1.2" />
  </ItemGroup>
  <ItemGroup>
    <Protobuf Include="../../../external/protos/concepts.proto" GrpcServices="Both" />
  </ItemGroup>
</Project>
```

```
> dotnet build
Microsoft (R) Build Engine version 16.5.0+d4cbfca49 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

CSC : error CS2001: Source file 'C:\Users\dev\Documents\solution\src\backend\project\../../../build/backend/debug-any-cpu/obj/ConceptsGrpc.cs' could not be found. [C:\Users\dev\Documents\solution\src\backend\project\project.csproj]

Build FAILED.
```

**Comment by jtattermusch** [comment]
Can someone provide a minimal example project that reproduces the problem? i.e. the csproj, the .proto files and all the other files needed and upload e.g. a github repository. Without it, it's hard to see what are all the options being used and how they affect the warning.
(it's not clear if the problem only happens in intermediate path is overriden, if the .proto file contains some services etc.).

Basically this could be due to the fact that Grpc.Tools "guesses" what will be the output files for given .proto (it could be File.cs and FileGrpc.cs, but FileGrpc.cs is not always generated, which could trigger the warning in theory - it's hard to know for sure without a concrete example one could experiment with)

**Comment by jtattermusch** [comment]
Btw, if you're just getting the warning but everything else works normally, things are likely working as expected.
You can also silence the warning, as described here:
https://github.com/grpc/grpc/blob/master/src/csharp/BUILD-INTEGRATION.md#i-am-getting-a-warning-about-a-missing-expected-file

Unless there's a proof that things are behaving differently than they should (as described in the docs link), I'm going to close.

**Comment by PetruRitivoiu** [comment]
Here's an example. Ignore the 2 errors, they happen because i did not commit the certificates. Everything works as expected. The warnings don't have any impact on project.

https://github.com/PetruRitivoiu/grpc-unary-basic

**Comment by 9ee1** [comment]
> Unless there's a proof that things are behaving differently than they should (as described in the docs link), I'm going to close.

My [comment ](https://github.com/grpc/grpc/issues/19816#issuecomment-616429373 )has an example of the build failing.

If the comment right above this one I am writing now with the example repository is insufficient, please let us know and I will be happy to create a test project and upload it for you.

**Comment by jtattermusch** [comment]
I've looked at https://github.com/PetruRitivoiu/grpc-unary-basic example and found what the problem was:

The  original project specifies this (not the slash at the end of OutputDir), which generates a warning
 `<Protobuf Include="proto\sqrt.proto" OutputDir="%(RelativePath)models/"  />`

When you specify this (no trailing slash in Output path), the warning goes away.
 `<Protobuf Include="proto\sqrt.proto" OutputDir="%(RelativePath)models"  />`

In MSBuild the trailing slash for path is not necessary (AFAIK including it is against MSBuild best practices) and removing it also removes the warning, so I'm going to assume that that's what's causing the problems.

**Comment by jtattermusch** [comment]
@9ee1  I also looked at your comment and the IntermediateOutputPath and OutputPath you are using have 2 problems
- you need to get rid of the trailing slash
- you need to use backslashes `\` in the paths instead of `/' (MSBuild uses backslashes for the paths it generates).

With these two fixes, the problem goes away.

**Comment by jeromv** [comment]
I actually have exactly the same problem. It gives a warning that the some expected protoc outputs were not generated. And then later my build fails because it needs these generated classes. I had them compiled to a folder "generated" changed that to OutputDir="%(RelativePath)generated" In Visual Studio all works fine

**Comment by 9ee1** [comment]
Thanks @jtattermusch - Your suggestions solved my problem.

**Comment by magmajia-pki** [comment]
@jtattermusch Great! Thanks!

**Comment by macrogreg** [comment]
It does not seem to be resolved for me. Created a new issue:
https://github.com/grpc/grpc/issues/29161
